### DISCUSSION TITLE
Is there a way to invoke InstallTool depending on the targets that will be run?

### DISCUSSION THREAD

**Discussion by paulomorgado** [cake-build/cake#4317] [Q&A] [answered]
Some builds require some expensive to acquire tools only for some targets.

It would be nice to have a way to filter that.

**Comment by nils-a** [comment]
While there is no built-in feature for it, it can be done.

Depending on the complexity of things, the "easy" way would be to try to split your script into multiple files, each file requiring only the tools needed for the tasks in that file. The "main" script file would then call the specific other files using [`CakeExecuteScript`](https://cakebuild.net/api/Cake.Common.Tools.Cake/CakeAliases/8E1BFC23).

Alternatively, it is also possible to add tools fully dynamically:

The idea here is, essentially, not to directly add the `#tool` directive to your script, but to create temporary file on-the-fly that contains that directive and then run that file also on-the-fly using [`CakeExecuteScript`](https://cakebuild.net/api/Cake.Common.Tools.Cake/CakeAliases/8E1BFC23).
This will neatly add the tool into the `tools` folder for later use.

[`Cake.Recipe`](https://github.com/cake-contrib/Cake.Recipe/) does exactly this in `RequireToolNotRegistered` / `RequireTool`: https://github.com/cake-contrib/Cake.Recipe/blob/6bee4e5cfbc2ad0a89483d360864da85c7d0bd71/Source/Cake.Recipe/Content/tools.cake

**Reply by paulomorgado** [reply]
Can the same be done for CakeFrosting?

**Reply by nils-a** [reply]
Hm. Not sure. 
@pascalberger, any ideas?

**Comment by pascalberger** [comment, marked as answer]
In Cake Frosting the `IToolInstaller` service can be used in a task:

```
[TaskName("MyTask")]
public sealed class MyTask : FrostingTask<BuildContext>
{
    private readonly IServiceProvider _provider;

    public MyTask(IServiceProvider provider)
    {
        _provider = provider;
    }

    public override void Run(BuildContext context)
    {
        var toolInstaller = (IToolInstaller)provider.GetService(typeof(IToolInstaller));
        toolInstaller.Install(new PackageReference("nuget:?package=MyPackage&version=1.0.0"));
    }
}
```
